File: 11-Bridge/src/go/tv.go
package main

import "fmt"

type Tv struct {
	enabled bool
	volume  int
	channel int
}

func NewTv() *Tv {
	return &Tv{enabled: false, volume: 30, channel: 1}
}

func (t *Tv) IsEnabled() bool { return t.enabled }

func (t *Tv) Enable() {
	t.enabled = true
	fmt.Println("  TV is ON")
}

func (t *Tv) Disable() {
	t.enabled = false
	fmt.Println("  TV is OFF")
}

func (t *Tv) GetVolume() int { return t.volume }

func (t *Tv) SetVolume(volume int) {
	if volume < 0 {
		volume = 0
	} else if volume > 100 {
		volume = 100
	}
	t.volume = volume
	fmt.Printf("  TV volume: %d\n", t.volume)
}

func (t *Tv) GetChannel() int { return t.channel }

func (t *Tv) SetChannel(channel int) {
	t.channel = channel
	fmt.Printf("  TV channel: %d\n", t.channel)
}

func (t *Tv) GetName() string { return "TV" }

File: 11-Bridge/src/rust/tv.rs
use crate::device::Device;

pub struct Tv {
    enabled: bool,
    volume: i32,
    channel: i32,
}

impl Tv {
    pub fn new() -> Self {
        Tv { enabled: false, volume: 30, channel: 1 }
    }
}

impl Device for Tv {
    fn is_enabled(&self) -> bool { self.enabled }

    fn enable(&mut self) {
        self.enabled = true;
        println!("  TV is ON");
    }

    fn disable(&mut self) {
        self.enabled = false;
        println!("  TV is OFF");
    }

    fn get_volume(&self) -> i32 { self.volume }

    fn set_volume(&mut self, volume: i32) {
        self.volume = volume.clamp(0, 100);
        println!("  TV volume: {}", self.volume);
    }

    fn get_channel(&self) -> i32 { self.channel }

    fn set_channel(&mut self, channel: i32) {
        self.channel = channel;
        println!("  TV channel: {}", self.channel);
    }

    fn get_name(&self) -> &str { "TV" }
}
